File: python/solutions/day02.py
from pathlib import Path

_pairs: list = []

def load():
    global _pairs
    day = Path(__file__).name.replace('.py', '')
    text = Path(f'inputs/{day}.txt').read_text()
    _pairs = list(map(lambda a: a.split('-'), text.split(',')))

def solveA():
    load()
    total = 0
    for pair in _pairs:
        start = int(pair[0])
        end = int(pair[1])
        max_len = len(pair[1])

        for i in range(2, max_len + 1, 2):
            h = int(i / 2)
            factor = 10**h + 1
            min_p = 10**(h - 1)
            max_p = 10**h - 1
            p_low = max(min_p, int((start + factor - 1) / factor))
            p_high = min(max_p, int(end / factor))
            if p_low > p_high:
                continue

            count = p_high - p_low + 1
            total += int(factor * (p_low + p_high) * count / 2)

    print(total)

def solveB():
    load()
    ids = set()
    for pair in _pairs:
        start = int(pair[0])
        end = int(pair[1])
        max_len = len(pair[1])
        for i in range(2, max_len + 1):
            for j in range(1, i):
                if i % j != 0:
                    continue
                r = int(i / j)
                if r < 2:
                    continue

                factor = 0
                term = 1
                for t in range(0, r):
                    factor += term
                    term *= 10**j

                min_p = 10**(j - 1)
                max_p = 10**j - 1
                p_low = max(min_p, int((start + factor - 1) / factor))
                p_high = min(max_p, int(end / factor))
                for p in range(p_low, p_high + 1):
                    ids.add(p * factor)

    total = 0
    for id_val in ids:
        total += id_val
    print(total)

File: rust/src/solutions/day02.rs
use std::cmp::{max, min};
use std::collections::HashSet;
use std::fs;

fn solve_a(pairs: Vec<(&str, &str)>) {
    let mut total :i64 = 0;
    for (pair_a, pair_b) in pairs {
        let start = pair_a.parse::<i64>().unwrap();
        let end = pair_b.parse::<i64>().unwrap();
        let max_len = pair_b.len() as u32;
        for i in (2..=max_len).step_by(2) {
            let h = i / 2;
            let factor = 10_i64.pow(h) + 1;
            let min_p = 10_i64.pow(h - 1);
            let max_p = 10_i64.pow(h) - 1;
            let p_low = max(min_p, (start + factor - 1) / factor);
            let p_high = min(max_p, end / factor);
            if p_low > p_high {
                continue;
            }

            let count = p_high - p_low + 1;
            total += factor * (p_low + p_high) * count / 2;
        }
    }
    println!("{}", total);
}

fn solve_b(pairs: Vec<(&str, &str)>) {
    let mut ids: HashSet<i64> = HashSet::new();
    for pair in pairs {
        let start = pair.0.parse::<i64>().unwrap();
        let end = pair.1.parse::<i64>().unwrap();
        let max_len = pair.1.len() as u32;
        for i in 2..=max_len {
            for j in 1..i {
                if i % j != 0 {
                    continue;
                }
                let r = i / j;
                if r < 2 {
                    continue;
                }

                let mut factor: i64 = 0;
                let mut term: i64 = 1;
                for _ in 0..r {
                    factor += term;
                    term *= 10_i64.pow(j as u32);
                }

                let min_p = 10_i64.pow(j - 1);
                let max_p = 10_i64.pow(j) - 1;
                let p_low = max(min_p, (start + factor - 1) / factor);
                let p_high = min(max_p, end / factor);
                for p in p_low..=p_high {
                    ids.insert(factor * p as i64);
                }
            }
        }
    }

    let mut total: i64 = 0;
    for id in ids.iter() {
        total += id;
    }
    println!("{:?}", total);
}

pub fn solve() {
    let text = fs::read_to_string("src/inputs/day02.txt")
        .unwrap();
    let pairs = text
        .split(',')
        .map(|s| s.split_once('-').unwrap())
        .collect::<Vec<(&str, &str)>>();
    let pairs_cpy = pairs.clone();
    solve_a(pairs);
    solve_b(pairs_cpy);
}